src/editor.rs

```diff
@@ -0,0 +1,182 @@
+    }
+
+    // The git binary path should be either in the `cmd` or `bin` subdirectory
+    // of the git-for-windows install path, while the `sh.exe` binary is located
+    // in the `bin` subdirectory.
+    Path::new(str::from_utf8(&output.stdout).ok()?.trim())
+        .canonicalize()
+        .ok()?
+        .parent()?
+        .parent()?
+        .join(r"bin\sh.exe")
+        .canonicalize()
+        .ok()
+}
+
+#[cfg(not(windows))]
+fn git_sh_path() -> Option<PathBuf> {
+    Some("/bin/sh".into())
+}
+
+/// Read the git configuration to determine the value for GIT_EDITOR.
+fn git_editor() -> Option<String> {
+    // Testing environment variable to force using the fallback editor instead
+    // of GIT_EDITOR.
+    if std::env::var("CARGO_VET_USE_FALLBACK_EDITOR").unwrap_or_default() == "1" {
+        return None;
+    }
+
+    let output = Command::new("git")
+        .arg("var")
+        .arg("GIT_EDITOR")
+        .output()
+        .ok()?;
+    if !output.status.success() {
+        return None;
+    }
+    Some(str::from_utf8(&output.stdout).ok()?.trim().to_owned())
+}
+
+#[cfg(windows)]
+const FALLBACK_EDITOR: &str = "notepad.exe";
+
+// NOTE: This is probably not as reliably available as `vi`, but is easier to
+// quit from for users who aren't familiar with vi.
+#[cfg(not(windows))]
+const FALLBACK_EDITOR: &str = "nano";
+
+/// Get a Command which can be used to invoke the user's EDITOR to edit a
+/// document when passed an argument. This will try to use the user's configured
+/// GIT_EDITOR when possible.
+pub fn editor_command() -> Command {
+    // Try to use the user's configured editor if we're able to locate their git
+    // install. If this fails, invoke the default editor instead.
+    //
+    // XXX: If we end up with commands which invoke the editor many times, it
+    // may eventually be worth adding some form of caching here.
+    match (git_sh_path(), git_editor()) {
+        (Some(git_sh), Some(git_editor)) => {
+            let mut cmd = Command::new(git_sh);
+            cmd.arg("-c")
+                .arg(format!("{} \"$@\"", git_editor))
+                .arg(git_editor);
+            return cmd;
+        }
+        (_, None) => {
+            warn!("Unable to determine user's GIT_EDITOR");
+        }
+        (None, Some(_)) => {
+            warn!("Unable to locate user's git install to invoke GIT_EDITOR");
+        }
+    }
+    warn!("Falling back to running '{}' directly", FALLBACK_EDITOR);
+    Command::new(FALLBACK_EDITOR)
+}
+
+/// Run the default editor configured through git (GIT_EDITOR) and use it to
+/// edit the given file path.
+pub fn run_editor(path: &Path) -> io::Result<ExitStatus> {
+    editor_command().arg(path).status()
+}
+
+// On windows some editors (notably notepad pre-windows 11) don't handle
+// unix line endings very well, so make sure to give them windows line
+// endings.
+#[cfg(windows)]
+const LINE_ENDING: &str = "\r\n";
+
+#[cfg(not(windows))]
+const LINE_ENDING: &str = "\n";
```

Comment:
Yes, I agree that git's work around newlines is very annoying. This was basically just added as a defensive measure for when opening the editor on Windows, as the windows newline has no risk of being preserved, and everything's normalized immediately after. If I didn't do this, I believe that the notepad fallback on older windows systems would break, because notepad didn't support unix line endings for the longest time.